dskafka/consumer.go

```diff
@@ -0,0 +1,553 @@
+func (c *Consumer) Close() error {
+	if c == nil || c.readers == nil {
+		return nil
+	}
+
+	var lastErr error
+	for _, reader := range c.readers {
+		if err := reader.Close(); err != nil {
+			log.Println("kafka: error closing reader:", err)
+			lastErr = err
+		}
+	}
+	return lastErr
+}
+
+// getOrCreateReader gets an existing reader or creates a new one for the specified topic and groupID.
+// Readers are cached using a key combining groupID and topic to allow multiple readers
+// for the same topic with different consumer groups.
+//
+// The reader is configured with the consumer's settings including:
+//   - Fetch sizes (MinBytes, MaxBytes)
+//   - Timeouts (MaxWait)
+//   - Offset management (CommitInterval, StartOffset)
+//   - Partition assignment (all partitions by default)
+//
+// If a specific partition is configured (>= 0), the reader will only read from that
+// partition and consumer group functionality is disabled.
+//
+// Parameters:
+//   - topic: The Kafka topic name to read from
+//   - groupID: Consumer group ID for coordination (empty for no consumer group)
+//
+// Returns:
+//   - *kafka.Reader: The reader instance for the topic/group combination
+//   - error: Any error that occurred during reader creation or offset setting
+func (c *Consumer) getOrCreateReader(topic, groupID string) (*kafka.Reader, error) {
+	if c == nil {
+		return nil, errors.New("kafka: consumer not initialized")
+	}
+
+	// Validate brokers configuration
+	if len(c.config.Brokers) == 0 {
+		return nil, errors.New("kafka: no brokers provided")
+	}
+
+	key := topic
+	if groupID != "" {
+		key = groupID + ":" + topic
+	}
+
+	if reader, exists := c.readers[key]; exists {
+		return reader, nil
+	}
+
+	// Create SASL mechanism for the reader
+	mechanism, err := scram.Mechanism(scram.SHA512, c.config.ClientCredentials.Username, c.config.ClientCredentials.Password)
+	if err != nil {
+		return nil, err
+	}
+
+	readerConfig := kafka.ReaderConfig{
+		Brokers:        c.config.Brokers,
+		Topic:          topic,
+		GroupID:        groupID,
+		MinBytes:       c.config.MinBytes,
+		MaxBytes:       c.config.MaxBytes,
+		MaxWait:        c.config.MaxWait,
+		CommitInterval: c.config.CommitInterval,
+		StartOffset:    c.config.StartOffset,
+		// Add SASL authentication transport
+		Dialer: &kafka.Dialer{
+			Timeout:       10 * time.Second,
+			DualStack:     true,
+			SASLMechanism: mechanism,
+		},
+	}
+
+	// Set partition if specified (for partition-specific reading)
+	if c.config.Partition >= 0 {
+		readerConfig.Partition = c.config.Partition
+		readerConfig.GroupID = "" // Can't use group ID with specific partition
+	}
+
+	reader := kafka.NewReader(readerConfig)
+
+	// Only set offset manually when NOT using consumer groups
+	// Consumer groups manage their own offsets automatically
+	if groupID == "" && c.config.Partition < 0 {
+		// Set manual offset for non-group consumers based on configuration
+		if err := reader.SetOffset(c.config.StartOffset); err != nil {
```

Comment:
Do we really need to setOffset for consumer groups? and even offset is required, it should be absolute value only, right? not sentinel number?